command/agent/auth/aws/aws.go

```diff
@@ -197,11 +249,60 @@ func (a *awsMethod) Authenticate(ctx context.Context, client *api.Client) (retTo
 }
 
 func (a *awsMethod) NewCreds() chan struct{} {
-	return nil
+	return a.credsFound
 }
 
-func (a *awsMethod) CredSuccess() {
-}
+func (a *awsMethod) CredSuccess() {}
 
 func (a *awsMethod) Shutdown() {
+	close(a.credsFound)
+	close(a.stopCh)
+}
+
+func (a *awsMethod) pollForCreds(accessKey, secretKey, sessionToken string, frequencySeconds int) {
+	ticker := time.NewTicker(time.Duration(frequencySeconds) * time.Second)
+	defer ticker.Stop()
+	for {
+		select {
+		case <-a.stopCh:
+			a.logger.Trace("shutdown triggered, stopping aws auth handler")
+			return
+		case <-ticker.C:
+			if err := a.checkCreds(accessKey, secretKey, sessionToken); err != nil {
+				a.logger.Warn("unable to retrieve current creds, retaining last creds", err)
+			}
+		}
+	}
+}
+
+func (a *awsMethod) checkCreds(accessKey, secretKey, sessionToken string) error {
+	a.credLock.Lock()
+	defer a.credLock.Unlock()
+
+	a.logger.Trace("checking for new credentials")
+	currentCreds, err := awsauth.RetrieveCreds(accessKey, secretKey, sessionToken)
+	if err != nil {
+		return err
+	}
+
+	currentVal, err := currentCreds.Get()
+	if err != nil {
+		return err
+	}
+	lastVal, err := a.lastCreds.Get()
+	if err != nil {
+		return err
+	}
+
+	// These will always have different pointers regardless of whether their
+	// values are identical, hence the use of DeepEqual.
+	if !a.lastCreds.IsExpired() && reflect.DeepEqual(currentVal, lastVal) {
```

Comment:
The creds types are just structs of strings, so you can avoid DeepEqual with: `*currentVal == *lastVal`.

Reply:
Actually, it looks like the AWS `Get()` method just returns a Value{}, not a pointer, so just `currentVal` == `lastVal` should work, I think.

Reply:
Well . . . I think that could cause a bug. The AliCloud Credential object is something in their library, not ours, and so the underlying structure could change any time we version bump of the AliCloud SDK, for either the auth or secrets plugin. That could change invisibly to us. 

DeepEqual recurses down the struct so if at any point arrays or child structs or maps are added, it will still do the right thing. I'd feel _much_ more comfortable with the code as-is.

If we changed how it works and a version bump broke equality here at some point, that would cause the agent to believe it was seeing new credentials every single time, which could cause unexpected behavior depending on the frequency at which it were set to poll for credentials.

Anyways, I think the benefit of DeepEqual outweighs any minor speed gain we might see from simply dereferencing the pointers.